sdks/python/apache_beam/io/hadoopfilesystem.py

```diff
@@ -0,0 +1,284 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+""":class:`~apache_beam.io.filesystem.FileSystem` implementation for accessing
+Hadoop Distributed File System files."""
+
+from __future__ import absolute_import
+
+import logging
+import posixpath
+import re
+
+from hdfs3 import HDFileSystem
+
+from apache_beam.io.filesystem import BeamIOError
+from apache_beam.io.filesystem import CompressedFile
+from apache_beam.io.filesystem import CompressionTypes
+from apache_beam.io.filesystem import FileMetadata
+from apache_beam.io.filesystem import FileSystem
+from apache_beam.io.filesystem import MatchResult
+
+__all__ = ['HadoopFileSystem']
+
+_HDFS_PREFIX = 'hdfs:/'
+_URL_RE = re.compile(r'^' + _HDFS_PREFIX + r'(/.*)')
+_COPY_BUFFER_SIZE = 2 ** 16
+
+
+# TODO(udim): Add @retry.with_exponential_backoff to some functions, like in
+# gcsio.py.
+
+
+class HadoopFileSystem(FileSystem):
+  """``FileSystem`` implementation that supports HDFS.
+
+  URL arguments to methods expect strings starting with ``hdfs://``.
+
+  Uses client library :class:`hdfs3.core.HDFileSystem`.
+  """
+
+  def __init__(self):
+    """Initialize a connection to HDFS.
+
+    Connection configuration is done using :doc:`hdfs`.
+    """
+    super(HadoopFileSystem, self).__init__()
+    self._hdfs_client = HDFileSystem()
+
+  @classmethod
```

Comment:
<!--thread_id:cc_157912894_t; commit:5e9211de2f83b679cfc3ea9e6769b1f22ac0215c; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Why are these classmethods ? Probably you should use @staticmethod since cls is never used.

<!--section:body-->
_parse_url changed to static, scheme left as is to be consistent with other implementations of the Filesystem base class.